shared/actions/pinentry.js

```diff
@@ -1,76 +1,84 @@
+    console.log('Asked for passphrase')
+    const {prompt, submitLabel, cancelLabel, windowTitle, retryLabel, features, type} = payload.pinentry
+    const {sessionID} = payload
 
-      const {prompt, submitLabel, cancelLabel, windowTitle, retryLabel, features, type} = payload.pinentry
-      const sessionID = payload.sessionID
+    // Stash response
+    sessionIDToResponse[String(sessionID)] = response
 
-      dispatch(
-        PinentryGen.createNewPinentry({
-          cancelLabel,
-          features,
-          prompt,
-          retryLabel,
-          sessionID,
-          submitLabel,
-          type,
-          windowTitle,
-        })
-      )
+    return [
+      PinentryGen.createNewPinentry({
+        cancelLabel,
+        prompt,
+        retryLabel,
+        sessionID,
+        showTyping: features.showTyping,
+        submitLabel,
+        type,
+        windowTitle,
+      }),
+    ]
+  })
+}
 
-      uglySessionIDResponseMapper[sessionID] = response
+function _onNewPinentry(action: PinentryGen.NewPinentryPayload) {
+  return Saga.put(
+    PinentryGen.createReplaceEntity({
+      entities: I.Map([[String(action.payload.sessionID), action.payload]]),
+      keyPath: ['sessionIDToPinentry'],
     })
-  }
+  )
 }
 
-export function onSubmit(
-  sessionID: number,
-  passphrase: string,
-  features: RPCTypes.GUIEntryFeatures
-): AsyncAction {
-  let result = {passphrase: passphrase}
-  for (const feature in features) {
-    result[feature] = features[feature]
-  }
-  return dispatch => {
-    dispatch(PinentryGen.createOnSubmit({sessionID}))
-    uglyResponse(sessionID, result)
-  }
+function _onSubmit(action: PinentryGen.OnSubmitPayload) {
+  const {sessionID, passphrase} = action.payload
+  _respond(sessionID, {passphrase, storeSecret: false})
+  return Saga.put(
+    PinentryGen.createDeleteEntity({
+      ids: [action.payload.sessionID],
+      keyPath: ['sessionIDToPinentry'],
+    })
+  )
 }
 
-export function onCancel(sessionID: number): AsyncAction {
-  return dispatch => {
-    dispatch(PinentryGen.createOnCancel({sessionID}))
-    uglyResponse(sessionID, null, {
-      code: RPCTypes.constantsStatusCode.scinputcanceled,
-      desc: 'Input canceled',
+function _onCancel(action: PinentryGen.OnCancelPayload) {
+  const {sessionID} = action.payload
+  _respond(sessionID, null, {
+    code: RPCTypes.constantsStatusCode.scinputcanceled,
+    desc: 'Input canceled',
+  })
+  return Saga.put(
+    PinentryGen.createDeleteEntity({
+      ids: [action.payload.sessionID],
+      keyPath: ['sessionIDToPinentry'],
     })
-  }
+  )
 }
 
-function uglyResponse(sessionID: number, result: any, err: ?any): void {
-  const response = uglySessionIDResponseMapper[sessionID]
+function _respond(sessionID: number, result: any, err: ?any): void {
+  const sessionKey = String(sessionID)
+  const response = sessionIDToResponse[sessionKey]
   if (response == null) {
     console.log('lost response reference')
```

Comment:
When the new logger stuff gets merged this should probably be logger.error becuase this should never happen